base_rest_abstract_attachment/services/abstract_attachable.py

```diff
@@ -0,0 +1,158 @@
+
+    _name = "abstract.attachable.service"
+    _inherit = "base.rest.service"
+    _description = __doc__
+
+    def _json_parser_attachments(self):
+        res = [
+            ("attachment_ids:attachments", ["id", "name"]),
+        ]
+        return res
+
+    def _subvalidator_attachments(self):
+        return {
+            "type": "list",
+            "schema": {
+                "type": "dict",
+                "schema": {"id": {"coerce": to_int}},
+            },
+        }
+
+    @restapi.method(
+        routes=[(["/<int:object_id>/attachments/<int:attachment_id>"], "GET")],
+        output_param=restapi.BinaryData(required=True),
+    )
+    def download_attachment(self, object_id=None, attachment_id=None):
+        record = self._get(object_id)
+        self._check_attachment_access(record)
+        attachment = self._get_attachment_for_record(record, attachment_id)
+        content = None
+        headers = [("X-Content-Type-Options", "nosniff")]
+        if attachment.type == "url":
+            r = requests.get(attachment.url)
+            headers.extend(
+                [
+                    (
+                        "Content-Type",
+                        r.headers.get("Content-Type", attachment.mimetype),
+                    ),
+                    (
+                        "Content-disposition",
+                        r.headers.get(
+                            "Content-Disposition", content_disposition(attachment.name)
+                        ),
+                    ),
+                    ("Content-Length", r.headers.get("Content-Length", 0)),
+                ]
+            )
+            content = BytesIO(r.content)
+        else:
+            headers.extend(
+                [
+                    ("Content-Type", attachment.mimetype),
+                    ("Content-Disposition", content_disposition(attachment.name)),
+                    ("Content-Length", len(attachment.raw)),
+                ]
+            )
+            content = attachment.raw
+        response = request.make_response(content, headers)
+        response.status_code = 200
+        return response
+
+    @restapi.method(
+        routes=[(["/<int:object_id>/attachments"], "POST")],
+        input_param=restapi.MultipartFormData(
+            {
+                "file": restapi.BinaryData(required=True),
+                "params": restapi.Datamodel("ir.attachment.input"),
+            }
+        ),
+        output_param=restapi.Datamodel("ir.attachment.output"),
+    )
+    def create_attachment(self, object_id=None, file=None, params=None):
+        record = self._get(object_id)
+        self._check_attachment_access(record)
+        vals = self._prepare_attachment_params(record, file, params.dump())
+        attachment = self.env["ir.attachment"].create(vals)
+        return self.env.datamodels["ir.attachment.output"].load(
+            attachment.jsonify(["id", "name"])[0]
```

Comment:
ouchhh odoo.Model -> dict -> DataModel -> json... Do you really need all these steps.. You could achieve the same result by using the orm mode on pydantic models... see (https://github.com/OCA/rest-framework/pull/221/files#diff-3a30a8db1624833eda6491b3f0eb7a0569c3c713b1031062ee1b0df895f9ab98R48)

Reply:
I've migrated to pydantic now